More correct form of /etc/hosts

regellosigkeitsaxiom opened the pull request:
Makes management of `/etc/hosts` more declarative by using `networking.hosts`.

###### Motivation for this change

It makes possible to point multiple domains to `localhost` in correct way: just adding `127.0.0.1 localhost.localdomain` to `networking.extraHosts` will not always work and is [incorrect](https://unix.stackexchange.com/questions/102660/hosts-file-is-it-incorrect-to-have-the-same-ip-address-on-multiple-lines).

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

regellosigkeitsaxiom committed:
Added networking.extraLocalHosts option

It adds its contents to '127.0.0.1' line of /etc/hosts
It makes possible to point multiple domains to localhost in correct way

0xABAB commented on nixos/modules/config/networking.nix:
```diff
@@ -187,11 +199,11 @@ in
 
         # /etc/hosts: Hostname-to-IP mappings.
-        "hosts".text =
+        "hosts".text = let foo = concatStringsSep " " cfg.extraLocalHosts; in
```
`foo`? Seriously?

Mic92 replied:
What my predecessor meant to say, is that `foo` is in general not a good name for a variable as it is too generic.

regellosigkeitsaxiom replied:
By naming this "variable" `foo` I meant that its name does not hold any significance in semantic space of whole program. It is just a technical variable for text that occurs in several (two in this case) places, an its purpose is minor increase in bug-pronness in a way that one will not change this text in one place and forget to do same in the other.

Its scope is limited to several lines, and its use is kinda obvious, so it should not be a problem (except greping source).

It is just a coding style thing, and I do not hold it dear. So if you propose how should be done, it should not be a problem at all.

0xABAB reviewed (changes requested):
Good initiative, but careless implementation.

joachifm commented:
Help me understand: it seems to me the current implementation ensures that the canonical localhost mapping comes first and so always will resolve correctly (unless you override the low-level `environment.etc."hosts".text` attr, but that defeats this PR as well).  The issue described in the linked discussion refers to situations where 127.0.0.1 resolves to something other than localhost (due to an earlier entry) but it's unclear how you can make that happen by adding stuff to `extraHosts` currently.

Can you elaborate?

joachifm commented:
Hrm, it occurred to me that you're probably talking about names other than localhost ...

volth commented:
Maybe just do not include ```127.0.0.1 localhost``` into "hosts".text if ```cfg.extraHosts``` contains "127.0.0.1" ?
Then ```cfg.extraHosts``` would have full control on the line and no need to introduce a new option

Mic92 commented on nixos/modules/config/networking.nix:
```diff
@@ -187,11 +199,11 @@ in
+        "hosts".text = let foo = concatStringsSep " " cfg.extraLocalHosts; in
           ''
-            127.0.0.1 localhost
+            127.0.0.1 localhost ${foo}
```
according to the article, it might be desirable to have the variable before localhost to set the fqdn correctly.

regellosigkeitsaxiom replied:
Thanks for correction, I will adjust it.

regellosigkeitsaxiom replied:
As for prepending `extraHosts` before localhost, this will cause havoc, as explained in chosen answer of article.

regellosigkeitsaxiom commented:
@joachifm my use case is the following.

I sometimes work on web services for which URI's are important, say it connects to socket foo.example.com. CORS things, web clients and stuff. When I work offline on my laptop, I start corresponding services locally and add them to 127.0.0.1 line in /etc/hosts. This way whole system behaves exactly like on live/development environment, and I can hack it offline and/or with very fast feedback.

My specific problem is, that if I am connected to network with no Internet connection, resolving second /etc/hosts 127.0.0.1 line hangs. I do not know exactly how this works, but two things help: setting down all external network adapters or setting up /etc/hosts correctly. Former is obviously a cruft, and the latter is correct one.

regellosigkeitsaxiom commented:
@volth I agree with you point that extra option for such trivial and rare thing is overkill.

However I think that doing it exactly the way you proposed will break config transparency. First, option is named `extraHosts` and given its type it is expected to append its content to config file verbatim. Handling line starting with 127.0.0.1 in a special way breaks this convention, makes it less transparent to user, needs parsing of raw text, which is easy both in implementation and introducing bugs, and surely is not the best thing given that Nix Expression is a typed language.

I suppose that this form should better implement you intention:
```Nix
{
  hosts = {
    "127.0.0.1" = [ "foo.example.com" ];
    "192.168.0.2"  = [ "fileserver.local" "nameserver.local" ];
  };
}
```
This way it can be trivially and safely merged into /etc/hosts in expected way.

volth commented:
> I suppose that this form should better implement you intention:
> ```nix
> {
>   networking.hosts = {
>     "127.0.0.1" = [ "foo.example.com" ];
>     "192.168.0.2" = [ "fileserver.local" "nameserver.local" ];
>   };
> }
> ```
> This way it can be trivially and safely merged into /etc/hosts in expected way.

Yes, I like this idea with merging.
A fileserver service could add to /etc/hosts the IPs it listens on, etc. Some NixOS services already have option like "which of the local IPs to listen", and it is not always 127.0.0.1

joachifm commented:
@regellosigkeitsaxiom would using a higher merge priority work for you? As in, specifying
```
extraHosts = mkBefore ''
  127.0.0.1 foo.bar.baz
  127.0.0.1 a.b.c
''
```
Then these entries should come immediately after the standard localhost mappings in the resulting hosts file. That'd accomplish the same thing as this new option, right?

regellosigkeitsaxiom commented:
@joachifm if I understood you correctly, resulting file will be:
```
127.0.0.1 foo.bar.baz
127.0.0.1 a.b.c
127.0.0.1 localhost
```
This will [break many things](https://lists.debian.org/debian-devel/2004/06/msg00468.html)

My aim is to have the file:
```
127.0.0.1 foo.bar.baz a.b.c. localhost
```

joachifm commented:
@regellosigkeitsaxiom no, the current implementation emits `127.0.0.1 localhost` before `extraHosts`.

regellosigkeitsaxiom committed:
Added networking.hosts and networking.fqdn options

regellosigkeitsaxiom commented:
Removed `networking.extraLocalHosts`
Added `networking.hosts` as described in [comment](https://github.com/NixOS/nixpkgs/pull/27105#issuecomment-313791494) in this thread.
Also added option `networking.fqdn`, just to be able to pin FQDN in one place.

0xABAB commented on nixos/modules/config/networking.nix:
```diff
@@ -20,12 +20,35 @@ in
+      default = null;
+      example = "foo.example.com";
+      description = ''
+        Full qualified domain name, if any.
```
Fully

0xABAB reviewed (approved):
I have no further comments after the details I mentioned have been addressed. 

Thanks!

regellosigkeitsaxiom committed:
Typo fix

Mic92 commented on nixos/modules/config/networking.nix:
```diff
@@ -188,11 +211,25 @@ in
+                then concatStringsSep " " ( filter (x : x != "localhost" ) ( getAttr "::1" cfg.hosts))
+                else "";
+              otherHosts = allToString ( removeAttrs cfg.hosts [ "127.0.0.1" "::1" ]);
+              maybeFQDN = if cfg.fqdn == null then "" else cfq.fqdn;
```
we have `lib.optionalString` for this.

Mic92 commented on nixos/modules/config/networking.nix:
```diff
@@ -188,11 +211,25 @@ in
+              userLocalHosts =
+                if builtins.hasAttr "127.0.0.1" cfg.hosts
+                then concatStringsSep " " ( filter (x : x != "localhost" ) ( getAttr "127.0.0.1" cfg.hosts))
+                else "";
```
ca be also shortened with `optionalString`.

Mic92 commented on nixos/modules/config/networking.nix:
```diff
@@ -188,11 +211,25 @@ in
+              userLocalHosts6 =
+                if builtins.hasAttr "::1" cfg.hosts
+                then concatStringsSep " " ( filter (x : x != "localhost" ) ( getAttr "::1" cfg.hosts))
+                else "";
```
same here.

regellosigkeitsaxiom committed:
Style adjustments

Also dangerous typo fix

regellosigkeitsaxiom committed:
Documentation fixes

volth commented on nixos/modules/config/networking.nix:
```diff
@@ -188,11 +211,23 @@ in
+                ( concatStringsSep " " ( filter (x : x != "localhost" ) ( getAttr "127.0.0.1" cfg.hosts)));
+              userLocalHosts6 = optionalString
+                ( builtins.hasAttr "::1" cfg.hosts )
+                ( concatStringsSep " " ( filter (x : x != "localhost" ) ( getAttr "::1" cfg.hosts)));
```
```concatStringsSep " " (lib.remove "localhost" cfg.hosts."::1")```

volth commented on nixos/modules/config/networking.nix:
```diff
@@ -188,11 +211,23 @@ in
+              allToString = set : concatStringsSep "\n" ( map ( oneToString set ) ( builtins.attrNames set ));
+              userLocalHosts = optionalString
+                ( builtins.hasAttr "127.0.0.1" cfg.hosts )
+                ( concatStringsSep " " ( filter (x : x != "localhost" ) ( getAttr "127.0.0.1" cfg.hosts)));
```
```concatStringsSep " " (lib.remove "localhost" cfg.hosts."127.0.0.1")```

volth commented on nixos/modules/config/networking.nix:
```diff
@@ -188,11 +211,23 @@ in
         # /etc/hosts: Hostname-to-IP mappings.
         "hosts".text =
+          let oneToString = set : ip : ip + " " + concatStringsSep " " ( getAttr ip set );
+              allToString = set : concatStringsSep "\n" ( map ( oneToString set ) ( builtins.attrNames set ));
```
```concatMapStringsSep "\n" (oneToString set) (builtins.attrNames set)```
or even without ```oneToString```:
```concatMapStringsSep "\n" (ip: "${ip} ${concatStringsSep set.${ip}}") (builtins.attrNames set)```

regellosigkeitsaxiom committed:
Style optimizations

regellosigkeitsaxiom committed:
Example of networking.hosts is now literalExample

vcunat commented:
It seems clear that having the same IP on multiple lines is bad.

I don't know these UNIX* details well, but do you have any idea why `hostname` and `domainname` is configured separately/independently of localhost's fully-qualified DNS name?  IMHO it feels messy when those differ.  If there's no good use case, perhaps we could link their configs in nixos, at least as defaults...

regellosigkeitsaxiom commented:
@vcunat will this configuration work for your purpose?

```Nix
{
  networking.hostname="foobar";
  networking.fqdn = cfg.networking.hostname + ".example.com.";
  networking.hosts = {
    "93.184.216.34" = [ cfg.netwroking.fqdn ];
  };
}
```

While it might be nice to have options like `networking.domain` and `networking.publicIP`, I have no idea what places beside /etc/hosts they should affect (and if nowadays they make any sense at all), so I myself would rather refrain from adding them until I gain this knowledge.

volth commented:
Another use case for mergeable ```/etc/hosts```: https://github.com/NixOS/nixpkgs/issues/24683#issuecomment-314631069

vcunat reviewed (approved):
Seems OK.

vcunat commented:
BTW, we do have `networking.domain` already.  I don't know of any particular "use cases", I was just wondering about the stuff.

volth commented:
ok, I cherry-picked it into my branch
The only issue so far is to find all spots with
```nix
   networking.extraHosts = ''
     127.0.0.1 example.com
   '';
```
and replace them with
```nix
  networking.hosts."127.0.0.1" = [ "example.com" ];
```

May be a warning should be added when ```networking.extraHosts``` contains something like "IP host" ? Many users have not seen this discussion.

volth commented on nixos/modules/config/networking.nix:
```diff
@@ -20,12 +20,35 @@ in
     networking.extraHosts = lib.mkOption {
       type = types.lines;
       default = "";
       example = "192.168.0.1 lanlocalhost";
```
It might have another example, because ```192.168.0.1 lanlocalhost``` is what is better to set via ```networking.hosts```
In case is there is no other meaningful example for  ```networking.extraHosts``` (except of plain forms of what is to be set in ```networking.hosts```), it might have sense to remove or deprecate this option.

regellosigkeitsaxiom replied:
I think that `networking.extraHosts` should be deprecated in future, `networking.hosts` is superior in every way.

I don't think that supplying different example, like `93.184.216.34 example.com` will make any difference, as it currently serves its purpose just fine.

volth replied:
I meant if /etc/hosts could contain something else except ip-domain pairs, then ```extraHosts```'s purpose would be to provide a way to include those extras into /etc/hosts and then the example should show one of those non-```ip-domain``` directive.

regellosigkeitsaxiom replied:
According to [manual](http://man7.org/linux/man-pages/man5/hosts.5.html), there are no other types of directives.

volth replied:
Then a deprecation warning like https://github.com/NixOS/nixpkgs/pull/3929/commits/05e1c29c6e2c99dcb7dc1ec5ab1d540707ff095d ?

volth commented on nixos/modules/config/networking.nix:
```diff
@@ -188,11 +211,23 @@ in
+              ::1 ${maybeFQDN} ${userLocalHosts6} localhost
             ''}
+            ${otherHosts}
             ${cfg.extraHosts}
```
Even if ```networking.extraHosts``` be kept for support legacy configs, shouldn't its content be parsed and merged with ```networking.hosts``` instead of be blindly appended to ```/etc/hosts```

regellosigkeitsaxiom replied:
I simply don't think that NixOS should parse anything in its config at all: we have types and config tree for that.

regellosigkeitsaxiom committed:
Deprecation warning for networking.extraHosts

volth commented:
Sorry for being so boring again with talking about small features. I am the only user of this nice feature, might be the only user ever as I cannot affect the acceptance of the pull request in any way.
<hr/>

There are two modes of creating files in ```/etc```: 
 * the default: create a symlink to a readonly file in ```/nix/store```
 * create a normal file which would be mutable, at least for ```root```

I think the latter is more convenient for ```/etc/hosts```, sometimes it is useful to add/remove a line in ```/etc/hosts``` without editing ```/etc/nixos/configuration.nix``` followed by ```nixos-rebuild```. It is like the ```mutableUsers``` feature.

So I'd write ```"hosts" = { mode = "0644"; text = ...; };``` instead of ```"hosts".text = ...;```

regellosigkeitsaxiom commented:
@volth will resulting `hosts` be overwritten with every `nixos-rebuild`? If so, it will silently destroy configurations and make system less predictable for forgetful user.

If it will be switched by something like `networking.hostsMutable` and will be `false` by default, it should be fine with most users. Also maybe if set to `true` then `hosts` should not be managed by nix at all.

I am against including it into this PR as it seems an entirely different feature to me.

volth commented:
> @volth will resulting hosts be overwritten with every nixos-rebuild? If so, it will silently destroy configurations and make system less predictable for forgetful user.

Yes, it is for ephemeral changes which are common in webmaster practice, as in your example to set ```you-dev-site.com``` to ```127.0.0.1``` sometime it is useful to set ```google.com``` to ```127.0.0.1``` to divert Chrome requests via mitm proxy, etc.

Goint on with password analogy, there are ```password``` and ```hashedPassword``` (do not mix them with ```initialPassword``` ```initialHashedPassword``` which overwrite ```/etc/{passwd,shadow}``` and reset password on ```nixos-rebuild``` destroying the password a mutable user set using command-line tools.

regellosigkeitsaxiom commented:
@volth Maybe something like `networking.hostsMutable = { "no" | "ephemeral" | "unmanaged" }`?

volth commented:
I usually do 
```sh
cd /etc
cp hosts hosts1
rm hosts           # remove symlink
cp hosts1 hosts    # how there is a regular file to edit
```

There are three such places where I convert symlinks to regular files:
1. ```/etc/hosts```
2. xml files in ```/etc/fonts```
3. ```/etc/nix/nix.conf``` (to set ```build-max-jobs = 1``` when ```nixos-rebuild``` fails and it is not clear why because of concurrent jobs logs to the same screen)

fpletz commented:
What is going on here? Are there comments by @volth missing? I see @regellosigkeitsaxiom answering him but he isn't a participant on this issue. Github Bug?

fpletz reviewed (approved):
Thanks! I welcome the change to a more declarative way to specify hosts. :+1:

volth commented:
@fpletz we were talking about making ```/etc/hosts``` a real file like ```/etc/passwd``` instead of a symlink to nix store

regellosigkeitsaxiom commented:
@fpletz actually there are more than ten comments missing, and not only from @volth. I have no idea what is going on.

vcunat commented:
I suppose `mutableHosts` mode could be added, but that seems a significantly larger change, and we most likely want to keep this static way as well.

volth commented:
@vcunat it is not about silly ```mutableHosts``` option and larger changes. All these is merely an misunderstanding which now has its own life. It was only about ```environment.etc."hosts".mode``` defaulting to ```"0644"``` instead of ```"symlink"```

NeQuissimus commented:
Oh, I am just catching up on this...

Can we not deprecate `extraHosts` and just keep it at the bottom of the `/etc/hosts` output?!
I like using https://github.com/StevenBlack/hosts and add lots of entries that way...

vcunat commented:
@volth: that would result into the file being writable (by root) but overwritten on each activation, right?  I wonder what's the use case &ndash; is there some SW that wants to write into it?

@NeQuissimus: right, adding a list from a (potentially large) file is something we *do* want to support.  I should've noticed.  Undeprecating `extraHosts` seems the best way ATM.

regellosigkeitsaxiom commented:
@NeQuissimus technically, extraHosts are working as always. Nix just emits a warning during rebuild. I submitted #27791 which removes this warning.

edolstra commented:
It's not exactly clear to me what this PR does to 127.0.0.1. If it adds a mapping for the hostname, then that's incorrect - that's done automatically via `nss_myhostname`.

volth commented:
> that would result into the file being writable (by root) but overwritten on each activation, right? I wonder what's the use case – is there some SW that wants to write into it?

@vcunat It is just a part of everyday life. The initial use case behind this PR was a webmaster with dev versions of sites on 127.0.0.1. Ok, lets watch a webmaster. Should adding a new website require nixos-rebuild? Should adding "google.com 127.0.0.1" to capture Chrome activity by mitm-proxy require nixos-rebuild? Even @NeQuissimus's case: should update of malware domain list require nixos-rebuild?

regellosigkeitsaxiom commented:
@edolstra it makes possible to concatenate all mappings to 127.0.0.1 into one line:
```
127.0.0.1 foo.bar localhost
```
Via `extraHosts` it can only be done like this:
```
127.0.0.1 localhost
127.0.0.1 foo.bar
```
which sometimes leads to problems as /etc/hosts actually works as two-way mapping. Thiis PR actually ensures that every IP occurs only once, and 127.0.0.1 is treated specially only in a way that it guarantees that it will map to `localhost`.

Regarding `networking.fqdn`: according to, for example, [this](https://serverfault.com/questions/331936/setting-the-hostname-fqdn-or-short-name), first column of /etc/hosts should always contain FQDN of the server. `networking.fqdn` always emits its contents in first column of 127.0.0.1. Since in "vanilla" linux it is done manually and such FQDN can differ from `hostname ++ domain name`, I added it as a separate option to support such use cases (however I do not know if those actually exist).

edolstra commented:
Mapping the FQDN to 127.0.0.1 is very very wrong. For example, consider a web server configuration that tells the server to listen on `${fqdn}`. Then it will now suddenly listen on 127.0.0.1, which is probably not what you want.

`nss_myhostname` addresses this by making the hostname map to some public IP, with 127.0.0.1 only as a fallback, IIRC.

regellosigkeitsaxiom commented:
@edolstra In a sense, it will not "suddenly" listen in 127.0.0.1 as `networking.fqdn` did not exist prior to this PR. But you point that it can lead to misunderstandings and cause a mess seems clear to me. Then should `networking.fqdn` be removed entirely? Or should something like `fqdnIP` added?

regellosigkeitsaxiom commented:
@volth in my case I go with `nixos-rebuild`. If I do not update channels, it takes only couple of seconds. Does you config includes something that depends of `/etc/hosts`?

If I were in you shoes, I'd try to include custom post-install script which flips `/etc/hosts` into regular file.

vcunat commented:
It's a fallback because `myhostname` is tried *after* (public) DNS.  https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/config/nsswitch.nix#L19

vcunat commented:
I do sometimes remove symlinks in `/etc/` and replace them by temporary changed copies, e.g. for `/etc/nix/machines`.  It seems to work OK; having nixos do copies instead of symlinks would make that slightly easier, though I'm not certain if it would be worth it.

edolstra commented:
@regellosigkeitsaxiom Well, prior to this PR, the FQDN would be looked up in DNS, which would return an address other than 127.0.0.1, assuming the FQDN exists in DNS.

BTW, the `hostname` manpage says:
> Technically: The FQDN is the name gethostbyname(2) returns for the host name returned by gethostname(2).
There is no mention of 127.0.0.1 there.

Possibly relevant: https://bugzilla.redhat.com/show_bug.cgi?id=530343

In any case, let's not mess with hostname mappings in /etc/hosts since it will have unpredictable results.

vcunat commented:
The `hostname(1)` manpage doesn't really ensure me in this matter:
> *-f, --fqdn, --long* Display  the  FQDN  (Fully Qualified Domain Name). A FQDN consists of a short host name and the DNS domain name. Unless you are using bind or NIS for host lookups you can change the FQDN and the DNS domain name (which is part of the FQDN) in the /etc/hosts file.

The related docs seem to mainly describe the implementation, IMHO, and not the (intended) meaning of it.  We can't significantly affect that mess, except perhaps building some clearer abstraction above it.

In any case, I believe we should *allow* changing the 127.0.0.1 + ::1 lines declaratively &ndash; adding more aliases (which is what this PR did), even though we may not be recommending to do so.  As noted, by default this PR changes nothing.  We might perhaps improve the descriptions to better explain what the options do.

*BTW, if you use `gethostbyaddress` on `127.0.0.1` or `::1`, I believe `/etc/hosts` will take priority and get the first entry for it, i.e. you should get either `localhost` or `networking.fqdn` (if you explicitly set that option).  My DNS intuition about the function would prefer returning the real FQDN of that machine, i.e. the name that public DNS would resolve to this machine's public IP, though getting back `localhost` might be more correct in some ways.*

aszlig commented on nixos/modules/config/networking.nix:
```diff
@@ -188,11 +214,23 @@ in
+                ( builtins.hasAttr "::1" cfg.hosts )
+                ( concatStringsSep " " ( remove "localhost" cfg.hosts."::1" ));
+              otherHosts = allToString ( removeAttrs cfg.hosts [ "127.0.0.1" "::1" ]);
+              maybeFQDN = optionalString ( cfg.fqdn != null ) cfg.fqdn;
```
Wouldn't it make more sense to put the following into `config` instead?

```nix
{
  networking.hosts = {
    "127.0.0.1" = [ "localhost" ];
  } // optionalAttrs cfg.enableIPv6 {
    "::1" = [ "localhost" ];
  };
}
```

That way the value in `environment.etc.hosts.text` can be reduced to:

```nix
let
  mkEntry = ip: hosts: "${ip} ${concatMapStringsSep " " hosts}";
in ''
  ${concatStringsSep "\n" (mapAttrsToList mkEntry cfg.hosts)}
  ${cfg.extraHosts}
''
```

vcunat replied:
The point of much of the fuss in this PR is that the first name after an address has a "special meaning" in `/etc/hosts`.

vcunat replied:
Still, as the `fqdn` option was removed, we might as well simplify the implementation...

aszlig replied:
@vcunat: We could still use `mkOrder 900` on `[ "localhost" ]` so that users can override it with `mkBefore` and `mkAfter`, or even go further and change the type of the `hosts` option to something like this:

```nix
types.attrsOf (submodule {
  options.canonical = mkOption { type = types.str; ... };
  options.aliases = mkOption { type = types.listOf types.str; ... };
})
```

vcunat replied:
Hmm, right, that would also be a possibility.

regellosigkeitsaxiom replied:
@aszlig its ability to merge worries me. IIRC, in case of merging two different configurations for same IP `canonical` will be overwritten, while it should be demoted to alias (with possible warning to user). It seems to be rare case, but still legal and should be predictable. Beside that, I like your variant.

aszlig replied:
@regellosigkeitsaxiom: A different way to handle this would be the other way around, mapping canonical hostnames to IP addresses. But either way, we can still provide our own merge function instead of the default one.

regellosigkeitsaxiom replied:
> we can still provide our own merge function instead of the default one.

This. Although I have no idea how to implement it.

> mapping canonical hostnames to IP addresses.

This is good too. Current implementation misses check that no hostname mapped to two different IP addresses, but this way it can be done trivially.

In my mind, it looks like this:
```Nix
{
  networking.hosts.aliases = {
    "foo.example.com" = "127.0.0.2";
    "bar.example.com" = "127.0.0.2";
  };
  networking.hosts.canonical = {
    "foobar.example.com" = "127.0.0.2";
  };
}
```
which will emit `127.0.0.2 foobar.example.com foo.example.com bar.example.com`.

aszlig replied:
I guess this is a bit verbose, but let me write an implementation which allows both ways (`host -> ip` *and* `ip -> host(s)`)...

aszlig replied:
Btw. this already should merge as expected:

```sh-session
$ nix-instantiate nixos --eval --strict --arg configuration '{ imports = [ { networking.hosts."127.0.0.1".canonical = "aaa"; } ]; networking.hosts."127.0.0.1".canonical = "bbb"; }' -A config.networking.hosts 
error: The unique option `networking.hosts.127.0.0.1.canonical' is defined multiple times, in `<unknown-file>' and `<unknown-file>'.
(use ‘--show-trace’ to show detailed location information)
$
```

volth replied:
Say I have ```mysql.example.com``` and ```nagios.example.com``` on the same IP.
And I do not care about result ```IP -> host```.
Why should I be forced to explicitly chose one of the names to be ```canonical```? And chose another one in case of if a "canonical" service shut down or migrated. With hundreds of IPs it is just crazy.

aszlig replied:
See #27804.

NeQuissimus commented:
Thank you @regellosigkeitsaxiom. Generally, deprecated things go away sooner or later, which is why I was concerned when I saw the warning :) We are all good now from what I can tell.

volth commented:
I am not sure it was intended but this seems strange (it does no harm though):

```

$ hostname
x220a

$ hostname -a
localhost

$ head /etc/hosts
127.0.0.1  example.com x220a localhost
::1  x220a localhost

```
